applications/IgaApplication/custom_elements/shell_3p_element.cpp

```diff
@@ -0,0 +1,766 @@
+
+        for (unsigned int i = 0; i < number_of_control_points; ++i) {
+            const array_1d<double, 3 >& acceleration = GetGeometry()[i].FastGetSolutionStepValue(ACCELERATION, Step);
+            const unsigned int index = i * 3;
+
+            rValues[index] = acceleration[0];
+            rValues[index + 1] = acceleration[1];
+            rValues[index + 2] = acceleration[2];
+        }
+    }
+
+    void Shell3pElement::EquationIdVector(
+        EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY;
+
+        const SizeType number_of_control_points = GetGeometry().size();
+
+        if (rResult.size() != 3 * number_of_control_points)
+            rResult.resize(3 * number_of_control_points, false);
+
+        const SizeType pos = this->GetGeometry()[0].GetDofPosition(DISPLACEMENT_X);
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) {
+            const SizeType index = i * 3;
+            rResult[index]     = GetGeometry()[i].GetDof(DISPLACEMENT_X, pos).EquationId();
+            rResult[index + 1] = GetGeometry()[i].GetDof(DISPLACEMENT_Y, pos + 1).EquationId();
+            rResult[index + 2] = GetGeometry()[i].GetDof(DISPLACEMENT_Z, pos + 2).EquationId();
+        }
+
+        KRATOS_CATCH("")
+    };
+
+    void Shell3pElement::GetDofList(
+        DofsVectorType& rElementalDofList,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY;
+
+        const SizeType number_of_control_points = GetGeometry().size();
+
+        rElementalDofList.resize(0);
+        rElementalDofList.reserve(3 * number_of_control_points);
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) {
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_X));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Y));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Z));
+        }
+
+        KRATOS_CATCH("")
+    };
+
+    ///@}
+    ///@name Check
+    ///@{
+
+    int Shell3pElement::Check(const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT)
+
+        // Verify that the constitutive law exists
+        if (this->GetProperties().Has(CONSTITUTIVE_LAW) == false)
+        {
+            KRATOS_ERROR << "Constitutive law not provided for property " << this->GetProperties().Id() << std::endl;
+        }
+        else
+        {
+            // Verify that the constitutive law has the correct dimension
+            KRATOS_ERROR_IF_NOT(this->GetProperties().Has(THICKNESS))
+                << "THICKNESS not provided for element " << this->Id() << std::endl;
+
+            // Check strain size
+            KRATOS_ERROR_IF_NOT(this->GetProperties().GetValue(CONSTITUTIVE_LAW)->GetStrainSize() == 3)
+                << "Wrong constitutive law used. This is a 2D element! Expected strain size is 3 (el id = ) "
```

Comment:
2D?

Reply:
It is a surface formulation and this requires a 2D constitutive law.